### ISSUE TITLE
Add `Backend.upsert`

### ISSUE THREAD

**Issue by jcrist** [ibis-project/ibis#5391]
Currently `ibis` supports `create_table` for creating (or replacing) a table, and `insert` for inserting new data into a table. Another common operation is `upsert`, which is supported by some backends. This is like `insert`, but also supports updating existing rows.

- For the data analysis DB tasks that `ibis` is targeted at is there a use case for this operation?
- What backends support this operation?
- If we add support, what options should `Backend.upsert` support?

**Comment by jcrist** [comment]
cc @prmoore77 - someone mentioned in passing that you might have some use cases for upsert support.

**Comment by cpcloud** [comment]
This might be useful, but let's wait until someone requests it.

**Comment by mw3i** [comment]
this would be a really nice feature to have. it would basically mean the only sql library my team needs to use is ibis (rather than the current awkward mix of sqlalchemy + dataset (for upserting) + ibis)

**Comment by lostmygithubaccount** [comment]
reopening for triage, thanks for the input @mw3i!

**Comment by prmoore77** [comment]
> cc @prmoore77 - someone mentioned in passing that you might have some use cases for upsert support.

hey @jcrist - yes, Upsert, or `MERGE` as it is known in quite a few databases is a very useful ETL tool.  Having that functionality in back-ends which support the `MERGE` DML statement would be very useful, and likely get more folks to use Ibis for ETL needs.  Thanks!

**Comment by lostmygithubaccount** [comment]
I think this and some related work in ensuring Ibis DDL things work w/ PyArrow tables and batches would be great and help close the loop. then you can run queries from one backend and use the results to populate tables in another, e.g. PySpark -> Postgres or Trino -> Clickhouse as we've heard recently

**Comment by NickCrews** [comment]
Thank you @mw3i, that looks great! Really if we implement #9186 and #9185, this would make it very easy to just basically pawn off this responsibility to users to do themselves with an external library. Same thing for other feature requests that come in: the easier we make it to interoperate with other libs, the less pressure there is for us to implement it ourselves.

**Comment by mw3i** [comment]
Fair point about interoperability (anything to reduce internal developer workload is a plus); though the dataset library in particular might be idiosyncratic (idk how many other libraries expect data to be formatted that way)

Given ibis leverages sqlalchemy already, it might not be too hard to maintain just 2 functions (insert and update, where update has an 'upsert' argument) that implement sqlalchemy inserts and upserts in the background without needing to make specific functions for each backend (though apologies if I'm being naive here)

I've seen some relative straightforward gists that implement [pseudo upserts in sqlalchemy](https://gist.github.com/malexer/0647b208b2f1c48ec93e1bd157dc67c0). Idk how easily something like that could fit into the framework ibis uses

**Comment by ramil-hx** [comment]
Can we implement Backend.upsert in Snowflake please. I need it for my project. I'm stuck how to insert and update if record exists.

**Comment by NickCrews** [comment]
@ramil-hx as a workaround, this is what I currently use to upsert records into a duckdb database.  To work with snowflake's dialect, you probably need to modify this slightly, both in regards to the SQL, and to check how they handle primary keys.

```python
from ibis.backends.sql import SQLBackend
from ibis.expr import types as ir

def upsert(db: SQLBackend, table_name: str, new_data: ir.Table) -> ir.Table:
    """Upsert new data into a table using duckdb's INSERT OR REPLACE INTO tbl.

    This assumes that the destination table has a primary key constraint.
    Otherwise, the new rows will just be appended.

    Not all fields need to be present in the new data.
    Missing fields will be set to NULL.

    Returns the appended table.
    """
    existing_schema = db.table(table_name).schema()
    new_data = new_data.cast(existing_schema)
    new_data = new_data.select(*existing_schema)
    new_sql = db.compile(new_data)
    sql = f"INSERT OR REPLACE INTO {table_name} BY NAME \n{new_sql};"
    db.raw_sql(sql)
    return db.table(table_name)
```

**Comment by lmores** [comment]
Hi @NickCrews (small world, eh!), have you ever used your workaround when `new_data` is a memtable? I get: `duckdb.duckdb.CatalogException: Catalog Error: Table with name ibis_pandas_memtable_yynerbs36bezdf3us5awk7grji does not exist!`.

In my use case the generated sql string is:

```sql
INSERT INTO target_table SELECT * FROM "ibis_pandas_memtable_yynerbs36bezdf3us5awk7grji" AS "t0" ON CONFLICT DO NOTHING;
```

Slightly different from yours as I need to run it both on duckdb and postgres, and needs to do nothing rather than updating; anyway, I totally see why the db engine complains! I am wondering if you managed to make it work also in these conditions.

**Comment by lmores** [comment]
Forgot to get back to you: you were right! My backend is postgres, creating an intermediate table as you suggested made it work.

**Comment by grv87** [comment]
Would full-fledged `merge` be supported, including `when not matched by source` and `delete`? Or this is the whole other issue?

**Comment by deepyaman** [comment]
> Would full-fledged `merge` be supported, including `when not matched by source` and `delete`? Or this is the whole other issue?

I think `Backend.upsert` should be implemented as a [SQLGlot `Merge`](https://github.com/tobymao/sqlglot/blob/v26.17.1/sqlglot/expressions.py#L6967-L6975), which does look like it supports things like `when`s. That said, I don't have much familiarity/opinions yet on whether it would make sense to expose this all; one of the maintainers probably has a clearer view.

BTW @cpcloud I have a couple integrations (Kedro, Dagster) that need (or would heavily benefit from) this just generic `Backend.upsert` support; does it sound fine to you to take a stab at this using SQLGlot `Merge` on the backend?

**Comment by deepyaman** [comment]
> Would full-fledged `merge` be supported, including `when not matched by source` and `delete`? Or this is the whole other issue?

@grv87 Having taken a stab at implementing `Backend.upsert()`, I think this should be a whole other issue. :) We can implement `Backend.upsert()` without defining a public API for `WHEN` expressions, which I assume would require more discussion. (FWIW I have looked into [how SQLframe supports a generic `merge` API](https://sqlframe.readthedocs.io/en/stable/bigquery/#merge-statement), and I assume there would be some similarity in terms of what would need to be exposed.)

**Comment by NickCrews** [comment]
Seems fine to me to stick with a simple .upsert() API for now, and add  a .merge_into() API in later if needed.
